glide-core/redis-rs/redis/src/cluster_async/mod.rs

```diff
@@ -639,26 +640,34 @@ fn route_for_pipeline(pipeline: &crate::Pipeline) -> RedisResult<Option<Route>>
         }
     }
 
-    // Find first specific slot and send to it. There's no need to check If later commands
-    // should be routed to a different slot, since the server will return an error indicating this.
-    pipeline.cmd_iter().map(route_for_command).try_fold(
-        None,
-        |chosen_route, next_cmd_route| match (chosen_route, next_cmd_route) {
-            (None, _) => Ok(next_cmd_route),
-            (_, None) => Ok(chosen_route),
-            (Some(chosen_route), Some(next_cmd_route)) => {
-                if chosen_route.slot() != next_cmd_route.slot() {
-                    Err((ErrorKind::CrossSlot, "Received crossed slots in pipeline").into())
-                } else if chosen_route.slot_addr() == SlotAddr::ReplicaOptional {
-                    Ok(Some(next_cmd_route))
-                } else {
-                    Ok(Some(chosen_route))
+    if pipeline.is_atomic() {
+        // Find first specific slot and send to it. There's no need to check If later commands
+        // should be routed to a different slot, since the server will return an error indicating this.
+        pipeline
+            .cmd_iter()
+            .map(route_for_command)
+            .try_fold(None, |chosen_route, next_cmd_route| {
+                match (chosen_route, next_cmd_route) {
+                    (None, _) => Ok(next_cmd_route),
+                    (_, None) => Ok(chosen_route),
+                    (Some(chosen_route), Some(next_cmd_route)) => {
+                        if chosen_route.slot() != next_cmd_route.slot() {
+                            Err((
+                                ErrorKind::CrossSlot,
+                                "Received crossed slots in transaction",
+                            )
+                                .into())
+                        } else {
+                            Ok(Some(chosen_route))
+                        }
+                    }
                 }
-            }
-        },
-    )
+            })
+    } else {
+        // Pipeline is not atomic, so we can have commands with different slots.
```

Comment:
Please add documentation to the function and state that for non atomic pipelines it returns none